python/private/py_twine.bzl

```diff
@@ -0,0 +1,106 @@
+"""Implementation of py_twine rule.
+
+Simply wraps the tool with a bash script and a runfiles manifest.
+See https://twine.readthedocs.io/
+"""
+
+load(":py_wheel.bzl", "PyWheelInfo")
+
+_attrs = {
+    "twine_bin": attr.label(
+        doc = """\
+A py_binary that runs the twine tool.
+The default value assumes you have `twine` listed in your own requirements.txt, and have run
+`pip_parse` with the default name of `pypi`.
+
+If these don't apply, you might use the `entry_point` helper to supply your own twine binary:
+```starlark
+load("@my_pip_parse_name//:requirements.bzl", "entry_point")
+py_twine(
+    ...
+    twine_bin = entry_point("twine"),
+)
+```
+
+Or of course you can supply a py_binary by some other means which is CLI-compatible with twine.
+
+Currently rules_python doesn't supply twine itself.
+Follow https://github.com/bazelbuild/rules_python/issues/1016
+""",
+        default = "@pypi_twine//:rules_python_wheel_entry_point_twine",
+        executable = True,
+        cfg = "exec",
+    ),
+    "wheel": attr.label(providers = [PyWheelInfo]),
+    "_runfiles_lib": attr.label(default = "@bazel_tools//tools/bash/runfiles"),
+}
+
+# Bash helper function for looking up runfiles.
+# Vendored from
+# https://github.com/bazelbuild/bazel/blob/master/tools/bash/runfiles/runfiles.bash
+BASH_RLOCATION_FUNCTION = r"""
+# --- begin runfiles.bash initialization v2 ---
+set -uo pipefail; f=bazel_tools/tools/bash/runfiles/runfiles.bash
+source "${RUNFILES_DIR:-/dev/null}/$f" 2>/dev/null || \
+source "$(grep -sm1 "^$f " "${RUNFILES_MANIFEST_FILE:-/dev/null}" | cut -f2- -d' ')" 2>/dev/null || \
+source "$0.runfiles/$f" 2>/dev/null || \
+source "$(grep -sm1 "^$f " "$0.runfiles_manifest" | cut -f2- -d' ')" 2>/dev/null || \
+source "$(grep -sm1 "^$f " "$0.exe.runfiles_manifest" | cut -f2- -d' ')" 2>/dev/null || \
+{ echo>&2 "ERROR: cannot find $f"; exit 1; }; f=; set -e
+# --- end runfiles.bash initialization v2 ---
+"""
+
+# Copied from https://github.com/aspect-build/bazel-lib/blob/main/lib/private/paths.bzl
+# to avoid a dependency from bazelbuild -> aspect-build
+def _to_manifest_path(ctx, file):
+    """The runfiles manifest entry path for a file.
+
+    This is the full runfiles path of a file including its workspace name as
+    the first segment. We refert to it as the manifest path as it is the path
+    flavor that is used for in the runfiles MANIFEST file.
+    We must avoid using non-normalized paths (workspace/../other_workspace/path)
+    in order to locate entries by their key.
+    Args:
```

Comment:
I'd avoid Bash for a few reasons.

1. It's not super portable, especially on Windows.
2. You end up using more subprocesses that are not Bash anymore.
3. The rlocation is a mess of code that could eventually break, and we keep copy-pasting it.

I'd suggest the following:

1. Rewrite it in Python (you can inherit the hermetic interpreter).
2. Ditch the rlocation for the expand_location in Starlark.

Reply:
For 1 this is true, I'd need a cmd.exe equivalent, we have that in lots of places.
For 2, I don't think a bash parent and python subprocess is any worse than two python processes. It might make a difference if we wanted to customize the twine CLI, but I don't want that.
For 3, I think it's pretty reliable, we could use the aspect_bazel_lib utils.bzl which exports that constant, but sadly I think we'll get in trouble using that from bazelbuild

Really what I want is not to have a parent process at all. It's a tiny annoying shim. How important do you think it is? I'm kinda out of time to work on this.

`expand_location` can't do the right thing if there's no runfiles tree, that's why the runfiles library needs to be used within the action. I think if it were a python parent process it would have to use the very runfiles library we are publishing here :)

Reply:
It would be funny to use the runfiles library to publish itself. :)

If you don't want to have parent processes at all, Python is the way to go because you can load twine and execute the code directly. Also, you don't need two implementations for multiple platforms, which adds maintenance complexity. I wonder how other maintainers feel about this.

Also, something that I forgot to suggest earlier is that you don't need to copy the wheel, you can symlink it to avoid duplication.

Reply:
I think what you want is to use `args`, so something like:

```
def py_package(name, ..):
  py_wheel(name=name, ...)
  run_twine(
    name=name + ".publish",
    args=['upload', "$(location :{wheel})".format(name)]
    data = [name]
  )

run_twine = rule(
  attrs={"_twine": attr.label(executable=True)}
)
def run_twine_impl(...): return DefaultInfo(executable=...)
```

Unfortunately, there isn't a good way to "embed" the args into the output executable in a particularly easy way.

Reply:
Yeah that's the problem, Bazel doesn't capture args or env for programs which aren't run as actions.

Reply:
The run command (and test command) respects `args` and `env` of the target it's given. Running as an action won't respect those. Passing one target to another won't automatically respect the inner's args/env.

Reply:
Yes that's right, I had it backwards. I'll play with this and see if we can avoid wrapping the twine binary